.github/actions/free-hosted-runner-disk-space/action.yml

```diff
@@ -1,18 +1,156 @@
+    required: false
+    default: 'true'
+  remove-boost:
+    description: 'Removes Boost C++ libraries. (frees ~1 GB)'
+    required: false
+    default: 'true'
+  remove-haskell:
+    description: 'Removes GHC (Haskell) artifacts. (frees ~2.7 GB)'
+    required: false
+    default: 'true'
+  remove-tools-cache:
+    description: 'Removes tools and dependencies cached. (frees ~3 GB)'
+    required: false
+    default: 'true'
+  remove-android:
+    description: 'Removes Android SDKs and Tools. (frees ~11 GB)'
+    required: false
+    default: 'false'
+  remove-codeql:
+    description: 'Removes CodeQL Action Bundles. (frees ~5.4 GB)'
+    required: false
+    default: 'false'
+  remove-docker-images:
+    description: 'Removes cached Docker images. (frees ~3 GB)'
+    required: false
+    default: 'false'
   to-remove:
-    description: Collection of files/folders to be removed
-    default: "/usr/share/dotnet /opt/ghc /usr/local/share/boost $AGENT_TOOLSDIRECTORY"
+    description: 'Additional collection of files/folders to be removed'
+    default: ''
     required: false
+
 runs:
-  using: composite
+  using: "composite"
   steps:
+    - name: Disk space report before modification
+      shell: bash
+      run: |
+        echo "Memory and swap:"
+        sudo free
+        echo
+        sudo swapon --show
+        echo
+
+        echo "Available storage:"
+        sudo df -h
+        echo
+
+    - name: Move /var/lib/docker/
+      if: ${{ inputs.only-remove-files == 'false' }}
+      run: sudo mv /var/lib/docker/ "${GITHUB_WORKSPACE}/docker"
+      shell: bash
+
     - name: Remove unnecessary files
       run: |
-        echo "Before cleanup"
-        df -h
-        sudo rm -rf ${{ inputs.to-remove }}
-        echo "After cleanup"
-        df -h
+        echo "Removing unwanted software... "
+        to_remove="${{ inputs.to-remove }}"
+        if [[ ${{ inputs.remove-dotnet }} == 'true' ]]; then
+          sudo rm -rf /usr/share/dotnet
+        fi
+        if [[ ${{ inputs.remove-android }} == 'true' ]]; then
+          sudo rm -rf /usr/local/lib/android
+        fi
+        if [[ ${{ inputs.remove-haskell }} == 'true' ]]; then
+          sudo rm -rf /opt/ghc
+        fi
+        if [[ ${{ inputs.remove-codeql }} == 'true' ]]; then
+          sudo rm -rf /opt/hostedtoolcache/CodeQL
+        fi
+        if [[ ${{ inputs.remove-docker-images }} == 'true' ]]; then
+          sudo docker image prune --all --force
+        fi
+        if [[ ${{ inputs.remove-boost }} == 'true' ]]; then
+          sudo rm -rf /usr/local/share/boost
+        fi
+        if [[ ${{ inputs.remove-tools-cache }} == 'true' ]]; then
+          sudo rm -rf $AGENT_TOOLSDIRECTORY
+        fi
+        if [[ -n "$to_remove" ]]; then
+          sudo rm -rf $to_remove
+        fi
+        echo "... done"
+      shell: bash
+
+    - name: Maximize build space
+      if: ${{ inputs.only-remove-files == 'false' }}
+      uses: easimon/maximize-build-space@v10
```

Comment:
Please use [SHA Pinning](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions), example [here](https://github.com/Alfresco/alfresco-build-tools/blob/master/.github/actions/calculate-next-internal-version/action.yml#L22).